rotkehlchen/chain/ethereum/modules/compound/v3/decoder.py

```diff
@@ -70,10 +101,151 @@ def decode_reward_claim(self, context: DecoderContext) -> DecodingOutput:
 
         return DEFAULT_DECODING_OUTPUT
 
+    def _decode_supply_or_withdraw_event(
+            self,
+            context: DecoderContext,
+            compound_token: EvmToken,
+    ) -> DecodingOutput:
+        """Decode a compound v3 supply or withdraw event. Takes decoder context and
+        the compound v3 wrapped token of the supplied/withdrawn underlying token"""
+        if (underlying_token := self.underlying_tokens.get(compound_token)) is None:
+            log.debug(
+                'At compound v3 supply/withdraw decoding of tx '
+                f'{context.transaction.tx_hash.hex()} the underlying token was not found.',
+            )
+            return DEFAULT_DECODING_OUTPUT
+
+        amount = asset_normalized_value(
+            amount=hex_or_bytes_to_int(context.tx_log.data),
+            asset=underlying_token,
+        )
+
+        paired_event = None
+        action_from_event_type, action_to_event_subtype, action_to_notes = None, None, None
+        for event in context.decoded_events:
+            if (
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.asset == underlying_token and amount == event.balance.amount
+            ):
+                if event.event_type == HistoryEventType.SPEND:
+                    event.event_type = HistoryEventType.DEPOSIT
+                    event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                    event.notes = f'Deposit {amount} {underlying_token.symbol} to compound v3'
+                    action_from_event_type = HistoryEventType.RECEIVE
+                    action_to_event_subtype = HistoryEventSubType.RECEIVE_WRAPPED
+                    action_to_notes = f'Receive {{amount}} {compound_token.symbol} from compound v3'  # {amount} to be replaced in post decoding  # noqa: E501
+                else:
+                    event.event_type = HistoryEventType.WITHDRAWAL
+                    event.event_subtype = HistoryEventSubType.REMOVE_ASSET
+                    event.notes = f'Withdraw {amount} {underlying_token.symbol} from compound v3'
+                    action_from_event_type = HistoryEventType.SPEND
+                    action_to_event_subtype = HistoryEventSubType.RETURN_WRAPPED
+                    action_to_notes = f'Return {amount} {compound_token.symbol} to compound v3'
+
+                event.counterparty = CPT_COMPOUND_V3
+                paired_event = event
+                break
+
+        if (
+            paired_event is None or
+            action_from_event_type is None or action_to_event_subtype is None
```

Comment:
as it is written with checking one of the two is enough but I guess it is good to have the check in case of future changes break the logic